cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -2257,6 +2381,21 @@ void RegisterHashAggregateBasic(FunctionRegistry* registry) {
     DCHECK_OK(func->AddKernel(MakeKernel(boolean(), HashAggregateInit<GroupedAllImpl>)));
     DCHECK_OK(registry->AddFunction(std::move(func)));
   }
+
+  {
+    auto func = std::make_shared<HashAggregateFunction>(
+        "hash_count_distinct", Arity::Binary(), &hash_count_distinct_doc);
+    DCHECK_OK(func->AddKernel(
+        MakeKernel(ValueDescr::ARRAY, HashAggregateInit<GroupedCountDistinctImpl>)));
```

Comment:
Nit: I'd like to avoid lazy initialization of the Grouper (just to keep it in `Kernel::init` because... that's the init method), should be doable by providing custom `init` for both of these

Reply:
Concretely: if a type is unsupported by Grouper then as currently written we won't detect that error until the first call to `consume`. Among other things, that means that an `Expression` would successfully bind

Reply:
Moved initialization into the custom Init function.